sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/LogicalRelation.scala

```diff
@@ -17,23 +17,35 @@
 package org.apache.spark.sql.execution.datasources
 
 import org.apache.spark.sql.catalyst.analysis.MultiInstanceRelation
-import org.apache.spark.sql.catalyst.expressions.{AttributeMap, AttributeReference}
+import org.apache.spark.sql.catalyst.expressions.{ExprId, AttributeMap, AttributeReference}
 import org.apache.spark.sql.catalyst.plans.logical.{LeafNode, LogicalPlan, Statistics}
 import org.apache.spark.sql.sources.BaseRelation
 
 /**
  * Used to link a [[BaseRelation]] in to a logical query plan.
+ *
+ * Note that sometimes we need to use `LogicalRelation` to replace an existing leaf node without
+ * changing the output attributes' IDs.  We can pass in the `expectedOutputIds` parameter for
+ * this purpose.
  */
-private[sql] case class LogicalRelation(relation: BaseRelation)
-  extends LeafNode
-  with MultiInstanceRelation {
+private[sql] final class LogicalRelation(val relation: BaseRelation, expectedOutputIds: Seq[ExprId])
```

Comment:
Had a discussion with @marmbrus and @cloud-fan, we will keep it as a case class and change other places where we need to use `unapply`. We will also pass in a `Seq[AttributeReference]` and do the work of adding the correct `exprId`s inside `LogicalRelation`.